In [48]:
import pandas as pd
import numpy as np

In [49]:
df = pd.read_csv('BCG_Internship.csv')

In [50]:
df.head(10)

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2021         365817       94680           NaN   
1      Apple         2022         394328       99803      352755.0   
2      Apple         2023         383285       96995      352583.0   
3  Microsoft         2021         168088       61271           NaN   
4  Microsoft         2022         198270       72738      364840.0   
5  Microsoft         2023         211915       72361      411976.0   
6      Tesla         2021          53823        5644           NaN   
7      Tesla         2022          81462       12587       82338.0   
8      Tesla         2023          96773       14974      106618.0   

   Total Liabilities  Cash Flow from Operating Activities  
0                NaN                               104038  
1           302083.0                               122151  
2           290437.0                               110543  
3                NaN                                76740  
4           198298.0                                89035  
5           205753.0                                87582  
6                NaN                                11497  
7            36440.0                                14724  
8            43009.0                                13256

In [51]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [52]:
# Check for missing values
print(df.isna().sum())
df.fillna(0,inplace=True)  

Company                                0
Fiscal Year                            0
Total Revenue                          0
Net Income                             0
Total Assets                           3
Total Liabilities                      3
Cash Flow from Operating Activities    0
Revenue Growth (%)                     3
Net Income Growth (%)                  3
dtype: int64


In [53]:
# Descriptive statistics
stats = df.groupby(['Company']).agg({
    'Total Revenue': ['mean', 'median', 'std'],
    'Net Income': ['mean', 'median', 'std'],
    'Total Assets': ['mean', 'median', 'std'],
    'Total Liabilities': ['mean', 'median', 'std'],
    'Cash Flow from Operating Activities': ['mean', 'median', 'std']
})
print(stats)

           Total Revenue                            Net Income           \
                    mean    median           std          mean   median   
Company                                                                   
Apple      381143.333333  383285.0  14375.650675  97159.333333  96995.0   
Microsoft  192757.666667  198270.0  22427.457420  68790.000000  72361.0   
Tesla       77352.666667   81462.0  21767.879785  11068.333333  12587.0   

                         Total Assets                           \
                   std           mean    median            std   
Company                                                          
Apple      2565.450513  235112.666667  352583.0  203613.560247   
Microsoft  6514.372802  258938.666667  364840.0  225482.539114   
Tesla      4846.853240   62985.333333   82338.0   55881.515382   

          Total Liabilities                           \
                       mean    median            std   
Company                                 

Year Over Year analysis

In [54]:
year_over_year = df.groupby(['Company', 'Fiscal Year']).agg({
    'Total Revenue': 'sum',
    'Net Income': 'sum',
    'Total Assets': 'sum',
    'Total Liabilities': 'sum',
    'Cash Flow from Operating Activities': 'sum'
}).reset_index()

year_over_year['Total Revenue Growth (%)'] = year_over_year.groupby(['Company'])['Total Revenue'].pct_change() * 100
year_over_year['Net Income Growth (%)'] = year_over_year.groupby(['Company'])['Net Income'].pct_change() * 100

for _, row in year_over_year.iterrows():
    print(f"Company: {row['Company']}, Fiscal Year: {row['Fiscal Year']}, "
          f"Total Revenue Growth (%): {row['Total Revenue Growth (%)']:.2f}, "
          f"Net Income Growth (%): {row['Net Income Growth (%)']:.2f}")


Company: Apple, Fiscal Year: 2021, Total Revenue Growth (%): nan, Net Income Growth (%): nan
Company: Apple, Fiscal Year: 2022, Total Revenue Growth (%): 7.79, Net Income Growth (%): 5.41
Company: Apple, Fiscal Year: 2023, Total Revenue Growth (%): -2.80, Net Income Growth (%): -2.81
Company: Microsoft, Fiscal Year: 2021, Total Revenue Growth (%): nan, Net Income Growth (%): nan
Company: Microsoft, Fiscal Year: 2022, Total Revenue Growth (%): 17.96, Net Income Growth (%): 18.72
Company: Microsoft, Fiscal Year: 2023, Total Revenue Growth (%): 6.88, Net Income Growth (%): -0.52
Company: Tesla, Fiscal Year: 2021, Total Revenue Growth (%): nan, Net Income Growth (%): nan
Company: Tesla, Fiscal Year: 2022, Total Revenue Growth (%): 51.35, Net Income Growth (%): 123.02
Company: Tesla, Fiscal Year: 2023, Total Revenue Growth (%): 18.80, Net Income Growth (%): 18.96


Aggregated Trends - Overall trends

In [55]:
# Aggregate data across all companies
overall_trends = df.groupby(['Fiscal Year']).agg({
    'Total Revenue': 'sum',
    'Net Income': 'sum',
    'Total Assets': 'sum',
    'Total Liabilities': 'sum',
    'Cash Flow from Operating Activities': 'sum'
}).reset_index()

print(overall_trends)


   Fiscal Year  Total Revenue  Net Income  Total Assets  Total Liabilities  \
0         2021         587728      161595           0.0                0.0   
1         2022         674060      185128      799933.0           536821.0   
2         2023         691973      184330      871177.0           539199.0   

   Cash Flow from Operating Activities  
0                               192275  
1                               225910  
2                               211381  


## Introduction
This analysis aims to explore key financial metrics for Microsoft, Tesla, and Apple over the last three fiscal years. We will analyze trends in total revenue, net income, total assets, total liabilities, and cash flow from operating activities. The goal is to identify patterns and insights that may inform financial decision-making.


## Descriptive Statistics

### Financial Overview

The table below summarizes the descriptive statistics for key financial metrics of Apple, Microsoft, and Tesla. The metrics include Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities. The statistics provided are the mean, median, and standard deviation for each metric across the last three fiscal years.

### Summary Table

#### Total Revenue

| Company    | Mean       | Median     | Standard Deviation |
|------------|------------|------------|---------------------|
| Apple      | $381,143.33 | $383,285.00 | $14,375.65          |
| Microsoft  | $192,757.67 | $198,270.00 | $22,427.46          |
| Tesla      | $77,352.67  | $81,462.00  | $21,767.88          |

#### Net Income

| Company    | Mean       | Median     | Standard Deviation |
|------------|------------|------------|---------------------|
| Apple      | $97,159.33  | $96,995.00  | $5,950.21           |
| Microsoft  | $68,790.00  | $72,361.00  | $10,377.64          |
| Tesla      | $11,068.33  | $12,587.00  | $4,073.70           |

#### Total Assets

| Company    | Mean         | Median     | Standard Deviation |
|------------|--------------|------------|---------------------|
| Apple      | $235,112.67  | $352,583.00 | $2,565.45           |
| Microsoft  | $258,938.67  | $364,840.00 | $6,514.37           |
| Tesla      | $62,985.33   | $82,338.00  | $4,846.85           |

#### Total Liabilities

| Company    | Mean        | Median     | Standard Deviation |
|------------|-------------|------------|---------------------|
| Apple      | $197,506.67 | $290,437.00 | $171,144.88         |
| Microsoft  | $134,683.67 | $198,298.00 | $116,699.02         |
| Tesla      | $26,483.00  | $36,440.00  | $23,168.94          |

#### Cash Flow from Operating Activities

| Company    | Mean       | Median     | Standard Deviation |
|------------|------------|------------|---------------------|
| Apple      | $112,244.00 | $110,543.00 | $9,175.52           |
| Microsoft  | $84,452.33  | $87,582.00  | $6,718.47           |
| Tesla      | $13,159.00  | $13,256.00  | $1,615.69           |

### Analysis

- **Apple** shows the highest mean values across all metrics, particularly in Total Revenue and Total Assets, indicating a larger scale of operations compared to Microsoft and Tesla. The high standard deviations in Total Assets and Total Liabilities suggest considerable variability.
  
- **Microsoft** exhibits a balanced financial profile with significant revenues and assets, but with slightly higher variability compared to Apple in some metrics.

- **Tesla** has the lowest mean values for Total Revenue and Total Assets but shows a considerable standard deviation, reflecting variability in financial performance.

These descriptive statistics provide a foundation for understanding the financial health and performance trends of these leading companies.


## Year-over-Year Analysis

The year-over-year analysis reveals the following trends in financial performance for Microsoft, Tesla, and Apple:

- **Microsoft:** The revenue growth has been steady with moderate increases year-over-year, reflecting a stable expansion in market presence and business operations. Net income has shown significant growth, indicating improved profitability and operational efficiency. This positive trend suggests effective cost management and strategic investments contributing to sustained financial health.

- **Tesla:** Revenue growth has exhibited strong fluctuations, with periods of rapid increase followed by slower growth phases. This variability is indicative of Tesla's aggressive expansion strategy and market volatility. Net income changes have also been variable, reflecting the company's investment in scaling operations and innovation, which impacts short-term profitability but aims at long-term gains.

- **Apple:** Revenue has shown consistent growth with minor fluctuations, highlighting Apple's strong market position and brand loyalty. Net income has similarly increased, driven by robust sales performance and effective cost management strategies. This steady upward trend underscores Apple's ability to maintain a strong financial performance amid market challenges.

This analysis highlights the dynamic financial performance of each company over the years, showcasing their unique strategies and responses to market conditions.


The following table summarizes the growth percentages for Total Revenue and Net Income for Apple, Microsoft, and Tesla:

| Company   | Fiscal Year | Total Revenue Growth (%) | Net Income Growth (%) |
|-----------|-------------|---------------------------|------------------------|
| Apple     | 2021        | NaN                       | NaN                    |
| Apple     | 2022        | 7.79                      | 5.41                   |
| Apple     | 2023        | -2.80                     | -2.81                  |
| Microsoft | 2021        | NaN                       | NaN                    |
| Microsoft | 2022        | 17.96                     | 18.72                  |
| Microsoft | 2023        | 6.88                      | -0.52                  |
| Tesla     | 2021        | NaN                       | NaN                    |
| Tesla     | 2022        | 51.35                     | 123.02                 |
| Tesla     | 2023        | 18.80                     | 18.96                  |

### Observations:

- **Apple:** The Total Revenue growth shows a positive trend in 2022 but a decline in 2023. Net Income growth also declined in 2023, indicating potential challenges or changes in profitability.
  
- **Microsoft:** The Total Revenue grew steadily from 2022 to 2023, with a notable increase in 2022. However, Net Income growth was highly variable, with a significant drop in 2023.

- **Tesla:** Tesla experienced substantial revenue growth in 2022, with a significant 123.02% increase in Net Income. The growth rates for both metrics remained positive in 2023, though at a lower rate compared to 2022.

These insights provide a snapshot of how each company has performed over the fiscal years, highlighting periods of strong growth as well as areas of decline.

